rust/src/shards/gui/widgets/text_input.rs

```diff
@@ -0,0 +1,309 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright © 2022 Fragcolor Pte. Ltd. */
+
+use super::TextInput;
+use crate::core::cloneVar;
+use crate::shard::Shard;
+use crate::shards::gui::util;
+use crate::shards::gui::EGUI_UI_SEQ_TYPE;
+use crate::shards::gui::PARENTS_UI_NAME;
+use crate::shards::gui::STRING_VAR_SLICE;
+use crate::shardsc;
+use crate::types::common_type;
+use crate::types::Context;
+use crate::types::ExposedInfo;
+use crate::types::ExposedTypes;
+use crate::types::InstanceData;
+use crate::types::ParamVar;
+use crate::types::Parameters;
+use crate::types::Type;
+use crate::types::Types;
+use crate::types::Var;
+use crate::types::NONE_TYPES;
+use crate::types::STRING_TYPES;
+use std::cmp::Ordering;
+use std::ffi::CStr;
+
+lazy_static! {
+  static ref TEXTINPUT_PARAMETERS: Parameters = vec![(
+    cstr!("Variable"),
+    cstr!("The variable that holds the input value."),
+    STRING_VAR_SLICE,
+  )
+    .into(),];
+}
+
+impl Default for TextInput {
+  fn default() -> Self {
+    let mut parents = ParamVar::default();
+    parents.set_name(PARENTS_UI_NAME);
+    Self {
+      parents,
+      requiring: Vec::new(),
+      variable: ParamVar::default(),
+      exposing: Vec::new(),
+      should_expose: false,
+      mutable_text: true,
+    }
+  }
+}
+
+impl AsRef<str> for Var {
+  fn as_ref(&self) -> &str {
+    if self.valueType != shardsc::SHType_String
+      && self.valueType != shardsc::SHType_Path
+      && self.valueType != shardsc::SHType_ContextVar
+      && self.valueType != shardsc::SHType_None
+    {
+      panic!("Expected None, String, Path or ContextVar variable, but casting failed.")
+    }
+
+    if self.valueType == shardsc::SHType_None {
+      return "";
+    }
+
+    unsafe {
+      CStr::from_ptr(
+        self.payload.__bindgen_anon_1.__bindgen_anon_2.stringValue as *mut std::os::raw::c_char,
+      )
+      .to_str()
+      .unwrap()
+    }
+  }
+}
+
+impl egui::TextBuffer for Var {
+  fn is_mutable(&self) -> bool {
+    true
+  }
+
+  fn insert_text(&mut self, text: &str, char_index: usize) -> usize {
```

Comment:
Could maybe clarify the variable names by clearly naming `text` => `inserted_text`/`inserted_text_length_` etc.

Reply:
Usually I like to keep the same name as defined in the trait to avoid confusion when reading the doc.

Reply:
It's a bit of a nitpick but it makes it harder to follow when you just bump into this function IMO, since you're dealing with 2 sets of pointers+lengths